File: src/main/java/cn/huangsy/FindTheLongestSubstring.java
package cn.huangsy;

import java.util.Arrays;

/**
 * @author huangsy
 * @date 2020/5/20 8:49
 */
public class FindTheLongestSubstring {

    public int findTheLongestSubstring(String s) {
        int[] x = new int[128];
        x['a'] = 1;
        x['e'] = 2;
        x['i'] = 4;
        x['o'] = 8;
        x['u'] = 16;

        int[] n = new int[32];
        Arrays.fill(n, -1);
        n[0] = 0;
        int m = 0,t = 0;
        char[] chars = s.toCharArray();
        for (int i = 0; i < chars.length; i++) {
            t ^= x[chars[i]];
            if (n[t] == -1) {
                n[t] = i + 1;
            } else {
                m = Math.max(m, i + 1 - n[t]);
            }

        }
        return m;
    }

    public static void main(String[] args) {
        FindTheLongestSubstring findTheLongestSubstring = new FindTheLongestSubstring();
        System.out.println(findTheLongestSubstring.findTheLongestSubstring("eleetminicoworoep"));
        System.out.println(findTheLongestSubstring.findTheLongestSubstring("leetcodeisgreat"));
        System.out.println(findTheLongestSubstring.findTheLongestSubstring("bcbcbc"));
    }

}

File: src/main/java/cn/leetcode/1371/FindTheLongestSubstring.kt
package cn.leetcode.`1371`

import cn.huangsy.ValidPalindrome
import java.util.*

/**
 * @author huangsy
 * @date 2020/5/20 8:45
 */
class FindTheLongestSubstring {

    fun findTheLongestSubstring(s: String): Int {
        val x = IntArray(128)
        x['a'.toInt()] = 1
        x['e'.toInt()] = 2
        x['i'.toInt()] = 4
        x['o'.toInt()] = 8
        x['u'.toInt()] = 16

        var chars = s.toCharArray()
        val n = IntArray(32)

        var m = 0
        var t = 0
        n[0] = 1

        for (i in 2 .. chars.size + 1) {
            t = t xor x[chars[i-2].toInt()]
            if (n[t] == 0) {
                n[t] = i
            } else {
                m = Math.max(m, i - n[t])
            }
        }
        return m
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val findTheLongestSubstring = FindTheLongestSubstring()
            println(findTheLongestSubstring.findTheLongestSubstring("eleetminicoworoep"))
            println(findTheLongestSubstring.findTheLongestSubstring("leetcodeisgreat"))
            println(findTheLongestSubstring.findTheLongestSubstring("bcbcbc"))
        }
    }
}
